### PR TITLE
Added support for macOS/Homebrew in `install-required-packages.sh`.

### PR DESCRIPTION
**Pull request by Ferroin**
##### Summary

This adds basic support for macOS with Homebrew to `install-required-packages.sh`.

It installs everything required to actually build Netdata (except Judy, which is not available through Homebrew).

It has two significant limitations compared to using it on Linux:

* It requires the user to have already installed bash via Homebrew. macOS only includes an ancient version of bash for licensing reasons, and the script requires bash 4.x or newer.
* It can't run as root. Homebrew explicitly refuses to run as EUID 0 for safety reasons.

##### Component Name

area/packaging

##### Description of testing that the developer performed

Verified operational on my local mac Mini running macOS Catalina.

##### Additional Information

This will preferentially use libraries provided directly by macOS to ones installed through Homebrew. It also forces Python 3 as Homebrew does not provide Python 2.7 (and it's a pain to support the version provided directly by macOS).

Fixes: #8282

### CODE REVIEW

File: packaging/installer/install-required-packages.sh [hunk 1]
```diff
@@ -11,6 +11,7 @@ renice 19 $$ > /dev/null 2> /dev/null
 ME="${0}"
 
 if [ "${BASH_VERSINFO[0]}" -lt "4" ]; then
+
```
Comment (prologic) [inline packaging/installer/install-required-packages.sh]:
```suggestion
```

File: packaging/installer/install-required-packages.sh [hunk 2]
```diff
@@ -1629,6 +1710,25 @@ install_swupd() {
...
+validate_install_brew() {
+  brew list | grep -q "${*}" || echo "${*}"
+}
+
+install_brew() {
+  # download the latest package info
+  if [ "${DRYRUN}" -eq 1 ]; then
+    echo >&2 " >> IMPORTANT << "
+    echo >&2 "    Please make sure your system is up to date"
+    echo >&2 "    by running:  ${sudo} brew upgrade "
```
Comment (prologic) [inline packaging/installer/install-required-packages.sh]:
```suggestion
    echo >&2 "    by running:  brew upgrade "
```

File: packaging/installer/install-required-packages.sh [hunk 3]
```diff
@@ -1629,6 +1710,25 @@ install_swupd() {
...
+    echo >&2
+  fi
+
+  run ${sudo} brew install "${@}"
```
Comment (prologic) [inline packaging/installer/install-required-packages.sh]:
```suggestion
  run brew install "${@}"
```

**Comment by Ferroin** [comment]
> I _think_ you need to drop the use of `sudo`? You don't normally run `brew` under `sudo` at all.

Indeed. Somehow I completely missed these.

**Comment by Ferroin** [comment]
@netdata/core Can someone in the core team who has a Mac they can test on check this and approve if it works please?

**Comment by prologic** [comment]
> @netdata/core Can someone in the core team who has a Mac they can test on check this and approve if it works please?

I will test this locally.

Also (_doesn't have to be in this PR_) but let's add this to our CI -- There is support for macOS runners in GHA See [Supported Environments](https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources)

**Comment by prologic** [comment]
Basic run works:

```sh
prologic@Jamess-iMac
Fri Mar 13 07:39:15
~/NetData/netdata
 (irp-homebrew) 0
$ ./packaging/installer/install-required-packages.sh --dont-wait
./packaging/installer/install-required-packages.sh: line 1130: validate_tree_macos: command not found

We detected these:
Distribution    : macos
Version         : 19.3.0
Codename        :
Package Manager : install_brew
Packages Tree   : macos
Detection Method: uname
Default Python v: 2

Searching for distro_sdk ...
Searching for autoconf_archive ...
Searching for autogen ...

The following command will be run:

 >> IMPORTANT <<
    Please make sure your system is up to date
    by running:  brew upgrade

brew install autoconf-archive autogen


brew install autoconf-archive autogen
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> New Formulae
hdt                                                   tlx                                                   xclogparser
==> Updated Formulae
arangodb          cheat             exploitdb         glances           kibana            radare2           rex               sceptre           txt2tags

==> Downloading https://homebrew.bintray.com/bottles/autoconf-archive-2019.01.06.catalina.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/f7/f793582f781bede0afc9b807ceaefe811114fb2248cf7fb87d770b8ded1bfdff?__gda__=exp=1584049903~hmac=eeedeae3deb8d73d
######################################################################## 100.0%
==> Pouring autoconf-archive-2019.01.06.catalina.bottle.tar.gz
🍺  /usr/local/Cellar/autoconf-archive/2019.01.06: 585 files, 3.4MB
==> Installing dependencies for autogen: bdw-gc and guile
==> Installing autogen dependency: bdw-gc
==> Downloading https://homebrew.bintray.com/bottles/bdw-gc-8.0.4.catalina.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/3c/3c8765da91b046c3ab3f73fb00f885608dc4d8415071ca0170f4aa7b1048a6bf?__gda__=exp=1584049906~hmac=40afe3d5548975bc
######################################################################## 100.0%
==> Pouring bdw-gc-8.0.4.catalina.bottle.tar.gz
🍺  /usr/local/Cellar/bdw-gc/8.0.4: 69 files, 1.5MB
==> Installing autogen dependency: guile
==> Downloading https://homebrew.bintray.com/bottles/guile-2.2.6.catalina.bottle.1.tar.gz
==> Downloading from https://akamai.bintray.com/7b/7bfe0c8f0f0ccbd3efa21dc515c149492d900e68e6bf6a802d4f9295c8bcf9e9?__gda__=exp=1584049909~hmac=299e81b4fdf1af7a
######################################################################## 100.0%
==> Pouring guile-2.2.6.catalina.bottle.1.tar.gz
🍺  /usr/local/Cellar/guile/2.2.6: 786 files, 51.5MB
==> Installing autogen
==> Downloading https://homebrew.bintray.com/bottles/autogen-5.18.16.catalina.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/e1/e103302688bc9f7e4493c8827133f3341ab701740249196ddbfc2f2cf4fc1246?__gda__=exp=1584049914~hmac=3f378030f88c96b3
######################################################################## 100.0%
==> Pouring autogen-5.18.16.catalina.bottle.tar.gz
🍺  /usr/local/Cellar/autogen/5.18.16: 101 files, 1.9MB

All Done! - Now proceed to the next step.
```

**Comment by prologic** [comment]
`netdata-all` argument fails:

```sh
prologic@Jamess-iMac
Fri Mar 13 07:39:56
~/NetData/netdata
 (irp-homebrew) 0
$ ./packaging/installer/install-required-packages.sh --dont-wait --non-interactive netdata-all
./packaging/installer/install-required-packages.sh: line 1130: validate_tree_macos: command not found

We detected these:
Distribution    : macos
Version         : 19.3.0
Codename        :
Package Manager : install_brew
Packages Tree   : macos
Detection Method: uname
Default Python v: 2

Searching for distro_sdk ...
Searching for autoconf_archive ...
Searching for libz_dev ...
Searching for libuuid_dev ...
Searching for libmnl_dev ...
Searching for lm_sensors ...
Searching for libuv ...
Searching for lz4 ...
Searching for openssl ...
Searching for judy ...
WARNING
package judy is not available in this system.
You may try to install without it.

Searching for python_pymongo ...
Searching for python_mysqldb ...
Searching for python_psycopg2 ...

The following command will be run:

 >> IMPORTANT <<
    Please make sure your system is up to date
    by running:  brew upgrade

brew install WAARNING\| python-mysql


brew install WAARNING\| python-mysql
Error: No available formula with the name "waarning|"
==> Searching for a previously deleted formula (in the last month)...
Warning: homebrew/core is shallow clone. To get complete history run:
  git -C "$(brew --repo homebrew/core)" fetch --unshallow

Error: No previously deleted formula found.
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
==> Searching taps...
==> Searching taps on GitHub...
Error: No formulae found in taps.



We are very sorry!

Installation of required packages failed.

What to do now:

  1. Make sure your system is updated.
     Most of the times, updating your system will resolve the issue.

  2. If the error message is about a specific package, try removing
     that package from the command and run it again.
     Depending on the broken package, you may be able to continue.

  3. Let us know. We may be able to help.
     Open a github issue with the above log, at:

           https://github.com/firehol/netdata/issues
```

**Comment by prologic** [comment]
`netdata` argument also fails similarly:

```sh
prologic@Jamess-iMac
Fri Mar 13 07:41:51
~/NetData/netdata
 (irp-homebrew) 1
$ ./packaging/installer/install-required-packages.sh --dont-wait --non-interactive netdata
./packaging/installer/install-required-packages.sh: line 1130: validate_tree_macos: command not found

We detected these:
Distribution    : macos
Version         : 19.3.0
Codename        :
Package Manager : install_brew
Packages Tree   : macos
Detection Method: uname
Default Python v: 2 (will install python3 too)

Searching for distro_sdk ...
Searching for autoconf_archive ...
Searching for libz_dev ...
Searching for libuuid_dev ...
Searching for libmnl_dev ...
Searching for libuv ...
Searching for lz4 ...
Searching for openssl ...
Searching for judy ...
WARNING
package judy is not available in this system.
You may try to install without it.

Searching for python3_pymongo ...

The following command will be run:

 >> IMPORTANT <<
    Please make sure your system is up to date
    by running:  brew upgrade

brew install WAARNING\|


brew install WAARNING\|
Error: No available formula with the name "waarning|"
==> Searching for a previously deleted formula (in the last month)...
Warning: homebrew/core is shallow clone. To get complete history run:
  git -C "$(brew --repo homebrew/core)" fetch --unshallow

Error: No previously deleted formula found.
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
==> Searching taps...
==> Searching taps on GitHub...
Error: No formulae found in taps.



We are very sorry!

Installation of required packages failed.

What to do now:

  1. Make sure your system is updated.
     Most of the times, updating your system will resolve the issue.

  2. If the error message is about a specific package, try removing
     that package from the command and run it again.
     Depending on the broken package, you may be able to continue.

  3. Let us know. We may be able to help.
     Open a github issue with the above log, at:

           https://github.com/firehol/netdata/issues
```

**Comment by prologic** [comment]
These lines seem a little weird to me?

> brew install WAARNING\|

**Comment by prologic** [comment]
Maybe its worth adding this to our CI in this PR afterall :)

**Comment by Ferroin** [comment]
> These lines seem a little weird to me?
> 
> > brew install WAARNING|

Yeah, that's due to a typo somewhere. I'll sort it out and add this to our CI.

**Comment by Ferroin** [comment]
Waiting on the merge of #8383 to add the macOS CI for this, as that's the place it would end up.

**Comment by prologic** [comment]
Is this ready?

**Comment by Ferroin** [comment]
Not quite, I kind fo forgot about it. I'll try to get the required CI added either today or Monday at which point it _should_ be ready.

**Comment by cosmix** [comment]
@Ferroin is this PR still relevant?

**Comment by Ferroin** [comment]
> @Ferroin is this PR still relevant?

Unless we've decided we don't want people using the kickstart installer on macOS, yes it's still relevant.

File: packaging/installer/install-required-packages.sh [hunk 4]
```diff
@@ -1011,6 +1084,7 @@ declare -A pkg_python3=(
   ['gentoo']="dev-lang/python"
   ['sabayon']="dev-lang/python:3.4"
   ['clearlinux']="python3-basic"
+  ['macos']="python"
```
Comment (cosmix) [inline packaging/installer/install-required-packages.sh:1087]: That should be `python3` (at least on Catalina.)

Comment (Ferroin) [reply packaging/installer/install-required-packages.sh:1087]: We're using Homebrew here, and the current formula for `python` pulls in Python 3.7.7 (see https://github.com/Homebrew/homebrew-core/blob/master/Formula/python.rb) independent of what macOS version is in use. This is also why the declaration a bit further up for `pkg_python` lists a warning for macOS, I haven't been able to figure out a way to get anything except Python 3 through Homebrew.

Comment (cosmix) [reply packaging/installer/install-required-packages.sh:1087]: It doesn't matter. Unversioned symlinks are only installed in `/usr/local/opt/python/libexec/bin` by default iirc, and running `python` on a prompt will typically resolved to the `2.7.x` binary.

Comment (Ferroin) [reply packaging/installer/install-required-packages.sh:1087]: That doesn't change what we need to pull in here. This isn't telling Netdata itself what to use, but telling the package manager what to install.

Actually getting the Python stuff working correctly on macOS is going to be more involved than just fixing the package installer script to pull in the right thing though, but fixing the package installer script to pull in the right thing will make it simpler to do.

Comment (cosmix) [reply packaging/installer/install-required-packages.sh:1087]: You're correct, my bad.

**Comment by krital** [comment]
What is the current status of this?

**Comment by Ferroin** [comment]
It’s merged and _should_ be working correctly.

**Pull request merged**